### ISSUE TITLE
Convenience Method to Find Force Field Parameters by Cosmetic Attributes

### ISSUE THREAD

**Issue by SimonBoothroyd** [openforcefield/openforcefield#210]
It would be helpful to have a convenience function on the force field object which allows us to find (or extract) force field parameters or nodes in general which have certain cosmetic attributes attached to them (such as `eval=True` or `compute_gradients=True`). @leeping I think this could be a useful feature for you also?

**Comment by davidlmobley** [comment]
Can you say a little bit about what you'd use this for? Just curious.

**Comment by jchodera** [comment]
@davidlmobley : We've been discussing how to extend the property calculator API to request gradients of certain parameters be computed. The current thinking is inspired by the API for machine learning frameworks like [pytorch](https://pytorch.org/), where the API allows a [`requires_grad` property](https://pytorch.org/docs/stable/notes/autograd.html) to be set to `True` for operations where we require a gradient.

One way to use this idea is to allow the SMIRNOFF parameters to have attributes that allow us to set which ones should have gradients computed, like
```python
# Tell the property estimator that we need the gradient for this parameter
forcefield.forces['vdW'].parameters['[#1:1]-[#7]'].compute_gradient = True
```
To make it easier to use this information in the property estimator, we could also use some convenience functions like
```python
# get a list of all parameters with compute_gradient = True from the `vdW` force handler
parameters_with_gradients = forcefield.forces['vdW'].parameters_with_gradients
# get a list of all parameters from all force handlers with compute_gradient = True
all_parameters_with_gradients = forcefield.parameters_with_gradients
```
We're still thinking about how we might be able to present views of the property estimator that allow us to make use of ecosystems like [tensorflow](https://www.tensorflow.org/) or [pytorch](https://pytorch.org/) that allow us to use these tools for model fitting and sampling of models, rather than having to engineer our own software to do this.

**Comment by davidlmobley** [comment]
Ah, OK. I didn't immediately see the connection between that discussion and this issue; thanks for clarifying.

**Comment by j-wags** [comment]
Pinging @proteneer who also wanted this feature. This isn't in the toolkit yet, but this code should effectively look things up by id or cosmetic attribute in the 0.2.0 toolkit:

```
search_id = 'a5'
for handler in ff._parameter_handlers.values():
    for param in handler.parameters:
        if param.id == search_id:
            print(param)
```
